src/sentry/debug_files/artifact_bundle_indexing.py

```diff
@@ -216,21 +230,81 @@ def mark_bundle_for_flat_file_indexing(
 
 @sentry_sdk.tracing.trace
 def remove_artifact_bundle_from_indexes(artifact_bundle: ArtifactBundle):
+    redis_client = get_redis_cluster_for_artifact_bundles()
+
     flat_file_indexes = ArtifactBundleFlatFileIndex.objects.filter(
         flatfileindexstate__artifact_bundle=artifact_bundle
     )
     for idx in flat_file_indexes:
-        identifier = FlatFileIdentifier(
-            project_id=idx.project_id, release=idx.release_name, dist=idx.dist_name
-        )
+        identifier = FlatFileIdentifier.from_index(idx)
 
         was_removed = update_artifact_bundle_index(
             identifier, bundles_to_remove=[artifact_bundle.id]
         )
         if not was_removed:
             metrics.incr("artifact_bundle_flat_file_indexing.removal.would_block")
-            # TODO: mark this for async removal
-            pass
+            redis_client.sadd(get_deletion_key(idx.id), artifact_bundle.id)
+
+
+@sentry_sdk.tracing.trace
+def backfill_artifact_index_updates():
+    redis_client = get_redis_cluster_for_artifact_bundles()
+
+    indexes_needing_update = list(
+        ArtifactBundleFlatFileIndex.objects.filter(
+            flatfileindexstate__indexing_state=ArtifactBundleIndexingState.NOT_INDEXED.value,
+        ).distinct()[:10]
+    )
+
+    # To avoid multiple backfill tasks from running into lock contention,
+    # we will randomize the order in which we update the indexes
+    random.shuffle(indexes_needing_update)
+
+    # First, we are processing all the indexes that need bundles *added* to them,
+    # we also process *removals* at the same time.
+    for index in indexes_needing_update:
+        identifier = FlatFileIdentifier.from_index(index)
+
+        artifact_bundles = ArtifactBundle.objects.filter(
+            flatfileindexstate__flat_file_index=index,
+            flatfileindexstate__indexing_state=ArtifactBundleIndexingState.NOT_INDEXED.value,
+        ).select_related("file")
+
+        bundles_to_add = []
+        for artifact_bundle in artifact_bundles:
+            with ArtifactBundleArchive(artifact_bundle.file.getfile()) as archive:
+                bundles_to_add.append(BundleManifest.from_artifact_bundle(artifact_bundle, archive))
+
+        deletion_key = get_deletion_key(index.id)
+        bundles_to_remove = [int(bundle_id) for bundle_id in redis_client.smembers(deletion_key)]
+
+        if bundles_to_add or bundles_to_remove:
+            update_artifact_bundle_index(
+                identifier,
+                blocking=True,
+                bundles_to_add=bundles_to_add,
+                bundles_to_remove=bundles_to_remove,
+            )
+            if bundles_to_remove:
+                redis_client.srem(deletion_key, *bundles_to_remove)
+
+    # Then, we process any pending removals marked in redis.
+    deletion_keys = redis_client.scan_iter(
```

Comment:
We need to be careful here, since we might need batching, to make the iteration more efficient if our redis instance has lots of keys.

Reply:
Ideally I would love to do the same randomization as the merges are doing to avoid lock contention. Also I honestly have no idea how to do batching here with redis? I believe it is doing batching / cursor-based pagination under the hood for this call, but not sure how to get more control over this.